File: admin/src/types/pin.ts
export type PinLevel = 'DONG' | 'GU' | 'CITY' | 'PROVINCE';

export interface Pin {
  id: string;
  name: string;
  slug: string;
  center: { lat: number; lng: number };
  boundary?: GeoJSON.Polygon;
  level: PinLevel;
  parentPinId: string | null;
  isActive: boolean;
  userCount: number;
  createdAt: string;
}

export interface PinStats {
  pinId: string;
  userCount: number;
  activeMatchRequests: number;
  weeklyGames: number;
  topRankedUsers: {
    id: string;
    nickname: string;
    tier: string;
    score: number;
  }[];
}

export interface PinPost {
  id: string;
  pinId: string;
  pin?: Pin;
  authorId: string;
  author?: {
    id: string;
    nickname: string;
    profileImageUrl: string | null;
  };
  title: string;
  content: string;
  category: 'GENERAL' | 'MATCH_SEEK' | 'REVIEW' | 'NOTICE';
  viewCount: number;
  likeCount: number;
  commentCount: number;
  isDeleted: boolean;
  isBlinded: boolean;
  images?: PostImage[];
  createdAt: string;
  updatedAt: string;
}

export interface PostImage {
  id: string;
  postId: string;
  imageUrl: string;
  sortOrder: number;
}

export interface Comment {
  id: string;
  postId: string;
  authorId: string;
  author?: {
    id: string;
    nickname: string;
    profileImageUrl: string | null;
  };
  parentId: string | null;
  content: string;
  isDeleted: boolean;
  isBlinded: boolean;
  replies?: Comment[];
  createdAt: string;
  updatedAt: string;
}

export interface PinListFilter {
  level?: PinLevel;
  isActive?: boolean;
  search?: string;
}

export interface PostListFilter {
  pinId?: string;
  category?: string;
  isDeleted?: boolean;
  isBlinded?: boolean;
  search?: string;
}

File: app/lib/models/pin.dart
/// 지역 핀 모델
class Pin {
  final String id;
  final String name;
  final String? slug;
  final double centerLatitude;
  final double centerLongitude;
  final String level; // DONG | GU | CITY | PROVINCE
  final String? parentPinId;
  final bool isActive;
  final int userCount;
  final int? activeMatchRequests;
  final DateTime createdAt;

  const Pin({
    required this.id,
    required this.name,
    this.slug,
    required this.centerLatitude,
    required this.centerLongitude,
    required this.level,
    this.parentPinId,
    this.isActive = true,
    required this.userCount,
    this.activeMatchRequests,
    required this.createdAt,
  });

  factory Pin.fromJson(Map<String, dynamic> json) {
    final center = json['center'] as Map<String, dynamic>?;
    return Pin(
      id: json['id'] as String,
      name: json['name'] as String,
      slug: json['slug'] as String?,
      centerLatitude: center != null
          ? (center['lat'] as num).toDouble()
          : (json['centerLatitude'] as num?)?.toDouble() ?? 0.0,
      centerLongitude: center != null
          ? (center['lng'] as num).toDouble()
          : (json['centerLongitude'] as num?)?.toDouble() ?? 0.0,
      level: json['level'] as String? ?? 'DONG',
      parentPinId: json['parentPinId'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      userCount: json['userCount'] as int? ?? 0,
      activeMatchRequests: json['activeMatchRequests'] as int?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'slug': slug,
        'center': {
          'lat': centerLatitude,
          'lng': centerLongitude,
        },
        'level': level,
        'parentPinId': parentPinId,
        'isActive': isActive,
        'userCount': userCount,
        'activeMatchRequests': activeMatchRequests,
        'createdAt': createdAt.toIso8601String(),
      };

  String get levelDisplayName {
    switch (level) {
      case 'DONG':
        return '동';
      case 'GU':
        return '구';
      case 'CITY':
        return '시';
      case 'PROVINCE':
        return '광역';
      default:
        return level;
    }
  }
}
